src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java

```diff
@@ -149,4 +153,68 @@ String computeCrc32c(String input) {
         return Long.toHexString(crc.getValue());
     }
 
+    private static final int NO_PROGRESS_MAX = 2;
+    private static final int SLEEP_MILLISECONDS = 1000;
+
+    <T, V> List<T> batchOperation(List<V> inputs, ApiFutureGenerator<T, V> generator) throws InterruptedException {
+        int inputSize = inputs.size();
+        // We drive the retry processing by which outputs have not been filled in,
+        // so we initialize the outputs to be all null -> not filled in.
+        List<T> outputs = new ArrayList<>(inputSize);
+        for (int i = 0; i < inputSize; i++) {
+            outputs.add(null);
+        }
+
+        int noProgressCount = 0;
+        while (true) {
+            List<ApiFuture<T>> futures = new ArrayList<>(inputSize);
+
+            // generate a request for every not completed output
+            int requestCount = 0;
+            for (int i = 0; i < inputSize; i++) {
+                if (outputs.get(i) != null) {
+                    futures.add(null);
+                } else {
+                    futures.add(generator.accept(inputs.get(i)));
+                    requestCount++;
+                }
+            }
+            if (requestCount == 0) {
+                break;
+            }
+
+            // tried to collect a response for every request we generated
+            int completeCount = 0;
+            for (int i = 0; i < inputSize; i++) {
+                ApiFuture<T> future = futures.get(i);
+                if (future != null) {
+                    try {
+                        outputs.set(i, future.get());
+                        completeCount++;
+                    } catch (DeadlineExceededException | UnavailableException ex) {
+                        logger.warn("Retry-able error in firestore future get - input: " +
+                            inputs.get(i) + " message: " + ex.getMessage());
+                    } catch (ExecutionException ex) {
+                        throw new FileSystemExecutionException("batch operation failed", ex);
+                    }
+                }
+            }
+            // If we completed our requests we are done
+            if (completeCount == requestCount) {
+                break;
+            }
+
```

Comment:
will there ever be a case where the requestCount is non-zero, but also not equal to the completeCount at this point in the loop?

Reply:
Yes, when errors are thrown from the get(), the completion count is not incremented. That is how we know we need to go around the loop and retry the incomplete requests.